### ISSUE TITLE
[BUG] simpleperf stat command event monitored % incorrect in per-core mode when profiling process

### ISSUE THREAD

**Issue by hardc0der** [android/ndk#1214]
#### Description

I'm using simpleperf built on AOSP master and when using the stat command to profile a process in conjunction with the `--per-core` option the percentage that shows up next to each event seems to be wrong. I'm using a command line like `simpleperf stat -p $(pidof myapp) -e cpu-cycles --per-core --duration 5 --log verbose` and it often shows less than 100% for the time the event was monitored on a specific CPU because of the way time_enabled and time_running are being accounted for. It ends up just using the values from the last thread processed for a given CPU.

#### Environment Details

Version: simpleperf built from AOSP master branch
Host OS: Ubuntu 16.04
ABI: 64-bit ARM / arm64-v8a

**Comment by yabincui** [comment]
Thanks for reporting it! I think I understand your problem. I can explain it below:
Normally, stat cmd opens one perf_event (or counter) for a thread. When the thread is migrated between cpus, the perf_event follows it. So we should see run_time_of_perf_event / enable_time_of_perf_event = 100%.
But with --per-core option, stat cmd opens one perf event for a thread on each cpu. When a thread is migrated between cpus, the perf_event on that cpu is in running state. So it's normal to see run_time_of_perf_event / enable_time_of_perf_event < 100%.

The problem is when run_time / enable_time < 100%, simpleperf can't distinguish it with hardware counter multiplexing caused by less hardware counters than perf events. So you may see info like
in https://android-review.googlesource.com/c/platform/system/extras/+/1220206/1/simpleperf/cmd_stat.cpp#1050.

The run_time and enable_time for a perf_event is written in the kernel, so simpleperf can't change it.
I will update doc in https://android.googlesource.com/platform/system/extras/+/master/simpleperf/doc/executable_commands_reference.md#show-event-count-per-core to explain it, and also comment in the stat cmd output.
Let me know if you have any more suggestions.

**Comment by yabinc** [comment]
Maybe the info printed at the end of `stat --per-core` already shows enough info:
simpleperf I cmd_stat.cpp:1050] A percentage represents runtime_of_a_thread_on_a_cpu / runtime_of_a_thread_on_all_cpus.
simpleperf I cmd_stat.cpp:1050] If percentage sum of a thread < 100%, or total event count for a running
simpleperf I cmd_stat.cpp:1050] thread is 0, probably because there are less hardware counters available than
simpleperf I cmd_stat.cpp:1050] events. Try --use-devfreq-counters if on a rooted device.

**Comment by hardc0der** [comment]
> Maybe the info printed at the end of `stat --per-core` already shows enough info:
> simpleperf I cmd_stat.cpp:1050] A percentage represents runtime_of_a_thread_on_a_cpu / runtime_of_a_thread_on_all_cpus.
> simpleperf I cmd_stat.cpp:1050] If percentage sum of a thread < 100%, or total event count for a running
> simpleperf I cmd_stat.cpp:1050] thread is 0, probably because there are less hardware counters available than
> simpleperf I cmd_stat.cpp:1050] events. Try --use-devfreq-counters if on a rooted device.

So one thing that I've experimented with is to make sure the run_time for a given event for a thread adds up to 100% of the enabled time for the thread across all cores. The info that gets spit out at the end kind of alludes to that but if you profile a process and only choose to aggregate `--per-core` then you don't get the information to allow you to manually do that. The other problem is that if you falsely report < 100 % that seems to throw off the GHz and cpus used metrics that are spit out since they are scaled based on run time.

**Comment by yabinc** [comment]
Sorry for the delay. I found a percent calculate bug while trying to reply, should be fixed in https://android-review.googlesource.com/c/platform/system/extras/+/1269422.

> make sure the run_time for a given event for a thread adds up to 100% of the enabled time for the thread across all cores. but if you profile a process and only choose to aggregate --per-core then you don't get the information to allow you to manually do that.

If you use --per-thread --per-core, simpleperf stat should report one entry for each thread running on each core. percentage sum for entries belong to the same thread should add up to 100% (can be 99% or 101% given the delta is 1%).
If you use --per-core, simpleperf stat should report one entry for all threads running on each core. percentage sum for all entries should add up to 100%. And the aggregated entries should belong to
one event type.
 
> The other problem is that if you falsely report < 100 % that seems to throw off the GHz and cpus used metrics that are spit out since they are scaled based on run time.

simpleperf reports percentage comment, but it doesn't scale counter values based on percentage.
The reason is using scale to fix problems caused by hardware counter multiplexing is buggy.

**Comment by hardc0der** [comment]
> Sorry for the delay. I found a percent calculate bug while trying to reply, should be fixed in https://android-review.googlesource.com/c/platform/system/extras/+/1269422.

Thanks, this does seem to work as I would have expected.

> > The other problem is that if you falsely report < 100 % that seems to throw off the GHz and cpus used metrics that are spit out since they are scaled based on run time.
> 
> simpleperf reports percentage comment, but it doesn't scale counter values based on percentage.
> The reason is using scale to fix problems caused by hardware counter multiplexing is buggy.

I was referring to the metrics that are computed like [GHz here](https://android.googlesource.com/platform/system/extras/+/master/simpleperf/cmd_stat.cpp#248). Another problem I noticed is that the running time used to compute the per-core metrics is the running time on an arbitrary CPU, not necessarily the CPU for which the metric is being computed: https://android.googlesource.com/platform/system/extras/+/master/simpleperf/cmd_stat.cpp#310

**Comment by yabinc** [comment]
No, the problems here happen after https://android-review.googlesource.com/c/platform/system/extras/+/1212765, which is Jan 23. And I haven't got a chance to update simpleperf prebuilt this year yet.

**Comment by DanAlbert** [comment]
Gotcha, thanks. If you're working on a fix already r22 sounds right then.

**Comment by yabinc** [comment]
Uploaded a fix https://android-review.googlesource.com/c/platform/system/extras/+/1274954/2. But haven't checked in yet. Please verify it if having time.

**Comment by hardc0der** [comment]
> Uploaded a fix https://android-review.googlesource.com/c/platform/system/extras/+/1274954/2. But haven't checked in yet. Please verify it if having time.

Thanks for this. I tried it out and the metrics report as expected now. One odd thing I've noticed even before is that when using the `--per-core` option when profiling a process the time_enabled is in some cases not the same for a given thread's event on each CPU so the percentages don't work out properly.  It even happens when just monitoring the `task-clock` event so lack of counters doesn't explain it. I primarily want to make sure I'm getting accurate profiling data which is what I use the percentages for.
